lib/collection/src/shards/local_shard/mod.rs

```diff
@@ -919,6 +920,61 @@ impl LocalShard {
     pub async fn shard_recovery_point(&self) -> RecoveryPoint {
         self.clock_map.lock().await.to_recovery_point()
     }
+
+    /// Resolve the WAL delta on the current shard for the given `recovery_point`
+    ///
+    /// The delta can be sent over to the node which the recovery point is from, to restore its
+    /// WAL making it consistent with the current shard.
+    ///
+    /// On success, a WAL record number from which the delta is resolved is returned.
+    /// If a WAL delta could not be resolved, an error is returned describing the failure.
+    pub async fn resolve_wal_delta(
+        &self,
+        mut recovery_point: RecoveryPoint,
+    ) -> Result<u64, WalDeltaError> {
+        if recovery_point.is_empty() {
+            return Err(WalDeltaError::Empty);
+        }
+
+        let last_seen_snapshot = self.clock_map.lock().await.to_recovery_point();
+
+        // If our current node has any lower last seen clock than the recovery point specifies,
+        // we cannot resolve a WAL delta
+        if recovery_point.has_any_higher(&last_seen_snapshot) {
+            return Err(WalDeltaError::HigherThanCurrent);
+        }
+
+        // Extend clock map with missing clocks this node know about
+        // Ensure the recovering node gets records for a clock it might not have seen yet
+        recovery_point.extend_with_missing_clocks(&last_seen_snapshot);
+
+        // Remove clocks that are equal to the current last seen
+        // We don't have to transfer any record for these
+        // TODO: do we want to remove higher clocks too, as the recovery node already has all data?
+        recovery_point.remove_equal_clocks(&last_seen_snapshot);
+
+        // TODO: check truncated clock values or each clock we have:
+        // TODO: - if truncated is higher, we cannot resolve diff
```

Comment:
We might want to merge with these TODOs in place, so we can resolve them later once a separate truncated clock map is implemented.